### PR TITLE
Ported gelu decomp to ref

### PR DESCRIPTION
**Pull request by Chillee**
Ugh... these are actually so painful to write without operator overloading lol. 

Decided to just utilize operator overloading, and xfail the ref tests for now.

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -455,7 +455,7 @@ def _distance(a, b):
 
             # TODO: consider adding some tolerance to this comparison
             msg = f"Reference result was farther ({ref_distance}) from the precise " \
-                  "computation than the torch result was ({torch_distance})!"
+                  f"computation than the torch result was ({torch_distance})!"
```
Comment (Chillee) [inline test/test_ops.py:458]: @ngimel and you said the automated static analysis wasn't helpful :^)

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -19497,6 +19504,15 @@ def __init__(
         "_refs.nn.functional.elu",
         torch_opinfo_name="nn.functional.elu",
     ),
+    PythonRefInfo(
+        "_refs.nn.functional.gelu",
+        torch_opinfo_name="nn.functional.gelu",
+        decorators=(
+            # Need FakeTensor support for meta coverage
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref_meta',),
+            DecorateInfo(unittest.expectedFailure, 'TestCommon', 'test_python_ref',),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:19513]: Wait why is this test failing?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -5425,6 +5425,12 @@ def sample_inputs_gelu(self, device, dtype, requires_grad, **kwargs):
                 kwargs=dict(approximate=approximate)))
     return tensors
 
+
+def error_inputs_gelu(op, device, **kwargs):
+    yield ErrorInput(SampleInput(make_tensor((), dtype=torch.float, device=device), kwargs={"approximate": "asdf"}),
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:5430]: Add a comment for what this error input is testing

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -14629,6 +14635,7 @@ def error_inputs_mean(op_info, device, **kwargs):
            aten_name="gelu",
            aten_backward_name='gelu_backward',
            ref=reference_gelu if TEST_SCIPY else _NOTHING,
+           error_inputs_func=error_inputs_gelu,
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:14638]: Nice addition

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -19497,6 +19504,15 @@ def __init__(
         "_refs.nn.functional.elu",
         torch_opinfo_name="nn.functional.elu",
     ),
+    PythonRefInfo(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:19507]: We should make this (and the gelu OpInfo) elementwise unary. The easiest way to do this is probably to model the original OpInfo with a variant for the tanh approximation (that would require splitting this OpInfo, too)

File: torch/_refs/nn/functional/__init__.py [hunk 6]
```diff
@@ -294,3 +294,34 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
...
+)
+def gelu(a: TensorLikeType, approximate: str = "none") -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.gelu
+    """
+    if not isinstance(a, TensorLike):
+        raise RuntimeError(
+            "Expected a tensor input for an elementwise unary operation!"
+        )
+    M_SQRT2 = 1.41421356237309504880
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:314]: we should probably create a constants.py, but it doesn't have to be in this PR

File: torch/_refs/nn/functional/__init__.py [hunk 7]
```diff
@@ -294,3 +294,34 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
...
+    M_SQRT1_2 = 0.70710678118654752440
+    M_2_SQRTPI = 1.12837916709551257390
+    if approximate == "tanh":
+        kBeta = M_SQRT2 * M_2_SQRTPI * 0.5
+        kKappa = 0.044715
+        x_cube = a * a * a
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:320]: nit: a_cube

File: torch/_refs/nn/functional/__init__.py [hunk 8]
```diff
@@ -294,3 +294,34 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
...
+
+
+@register_decomposition(torch.ops.aten.gelu)
+@out_wrapper
+@elementwise_unary_scalar_wrapper
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a",),
+    type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def gelu(a: TensorLikeType, approximate: str = "none") -> TensorLikeType:
```
Comment (mruberry) [inline torch/_refs/nn/functional/__init__.py:306]: cool decomp

File: test/test_ops.py (hunk 1 above)
Comment (mruberry) [reply test/test_ops.py:458]: nice fix

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (Chillee) [reply torch/testing/_internal/common_methods_invocations.py:19513]: Currently we don't rebind `torch.Tensor.mul` to `_refs.mul`. It should pass after that.

I originally wrote this with `torch.mul` to make it pass, but it's actually just too ugly to handle.

Comment (Chillee) [reply torch/testing/_internal/common_methods_invocations.py:19513]: Actually, don't see any reason why it isn't done today. I added it to the PR. (cc: @ezyang )

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (Chillee) [reply torch/testing/_internal/common_methods_invocations.py:19507]: Can we do this in a later PR? This looks a little tricky.

**Review by mruberry — APPROVED** [review]
Approving for velocity but this needs a rebase and we should follow-up by updating the tests for gelu appropriately

**Comment by Chillee** [comment]
I added a todo in the opinfo.

**Pull request closed without merging**